File: 121. Best Time to Buy and Sell Stock/maxProfit.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    var buy = prices[0]
    var profit = 0
    prices.forEach(function(p){
        buy = Math.min(buy,p)
        profit = Math.max(profit, p-buy)
    });
    return profit
};

// Runtime 107 ms / Memory 51.8 MB
// debug: https://jsfiddle.net/

File: 121. Best Time to Buy and Sell Stock/maxProfit.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        int buy = prices[0];
        // DP
        for (auto price:prices) {
            if (buy > price) { buy = price; } //low cost
            profit = max(profit, price-buy); //best profit
        }

        return profit;
    }
};